btree/btree.c

```diff
@@ -0,0 +1,13037 @@
+				     m0_bcount_t vsize,
+				     struct m0_be_tx_credit *accum);
+
+	/**
+	 * Calculates credits required to delete the record and adds those
+	 * credits to @accum.
+	 */
+	void (*nt_rec_del_credit)(const struct nd *node, m0_bcount_t ksize,
+				  m0_bcount_t vsize,
+				  struct m0_be_tx_credit *accum);
+
+};
+
+/**
+ * Node descriptor.
+ *
+ * This structure is allocated (outside of the segment) for each node actively
+ * used by the b-tree module. Node descriptors are cached.
+ */
+struct nd {
+	struct segaddr          n_addr;
+	struct td              *n_tree;
+	const struct node_type *n_type;
+
+	/**
+	 * BE segment address is needed for LRU nodes because we set tree
+	 * descriptor to NULL and therefore loose access to segment information.
+	 */
+	struct m0_be_seg       *n_seg;
+	/**
+	 * Linkage into node descriptor list.
+	 * ndlist_tl, btree_active_nds, btree_lru_nds.
+	 */
+	struct m0_tlink	        n_linkage;
+	uint64_t                n_magic;
+
+	/**
+	 * The lock that protects the fields below. The fields above are
+	 * read-only after the node is loaded into memory.
+	 */
+	struct m0_rwlock        n_lock;
+
+	/**
+	 * Node refernce count. n_ref count indicates the number of times this
+	 * node is fetched for different operations (KV delete, put, get etc.).
+	 * If the n_ref count is non-zero the node should be in active node
+	 * descriptor list. Once n_ref count reaches, it means the node is not
+	 * in use by any operation and is safe to move to global lru list.
+	 */
+	int                     n_ref;
+
+	/**
+	 * Transaction reference count. A non-zero txref value indicates
+	 * the active transactions for this node. Once the txref count goes to
+	 * '0' the segment data in the mmapped memory can be released if the
+	 * kernel starts to run out of physical memory in the system.
+	 */
+	int                     n_txref;
+
+	uint64_t                n_seq;
+	struct node_op         *n_op;
+
+	/**
+	 *  Size of the node on BE segment pointed by this nd. Added here for
+	 *  easy reference.
+	 */
+	uint32_t                n_size;
+
+	/**
+	 * flag for indicating if node on BE segment is valid or not, but it
+	 * does not indicated anything about node descriptor validity.
+	 */
+	bool                    n_be_node_valid;
+};
+
+enum node_opcode {
+	NOP_LOAD = 1,
+	NOP_ALLOC,
+	NOP_FREE
+};
+
+/**
+ * Node operation state-machine.
+ *
+ * This represents a state-machine used to execute a potentially blocking tree
+ * or node operation.
+ */
+struct node_op {
+	/** Operation to do. */
+	enum node_opcode no_opc;
+	struct m0_sm_op  no_op;
+
+	/** Which tree to operate on. */
+	struct td       *no_tree;
+
+	/** Address of the node withing the segment. */
+	struct segaddr   no_addr;
+
+	/** The node to operate on. */
+	struct nd       *no_node;
+
+	/** Optional transaction. */
+	struct m0_be_tx *no_tx;
+
+	/** Next operation acting on the same node. */
+	struct node_op  *no_next;
```

Comment:
Is this field used ?

Reply:
No. I think they were introduced but found no use. I have deleted them.